source/common/upstream/cluster_manager_impl.cc

```diff
@@ -1027,6 +1028,194 @@ ClusterManagerImpl::addThreadLocalClusterUpdateCallbacks(ClusterUpdateCallbacks&
+    parent_.thread_local_dispatcher_.post([this] {
+      // Something might got added in the meantime, so check the map again.
+      if (pending_clusters_.empty()) {
+        callbacks_handle_.reset();
+        callbacks_.reset();
+      }
+      callbacks_handle_cleanup_posted_ = false;
+    });
+    callbacks_handle_cleanup_posted_ = true;
+  }
+}
+
+OdCdsApiHandleSharedPtr
+ClusterManagerImpl::allocateOdCdsApi(const envoy::config::core::v3::ConfigSource& odcds_config,
+                                     OptRef<xds::core::v3::ResourceLocator> odcds_resources_locator,
+                                     ProtobufMessage::ValidationVisitor& validation_visitor) {
+  // TODO(krnowak): Instead of creating a new handle every time, store the handles internally and
+  // return an already existing one if the config or locator matches. Note that this may need a way
+  // to clean up the unused handles, so we can close the unnecessary connections.
+  auto odcds = OdCdsApiImpl::create(odcds_config, odcds_resources_locator, *this, stats_,
+                                    validation_visitor);
+  return OdCdsApiHandleImpl::create(*this, std::move(odcds));
+}
+
+ClusterDiscoveryCallbackHandlePtr
+ClusterManagerImpl::requestOnDemandClusterDiscovery(OdCdsApiHandleImplSharedPtr odcds_handle,
+                                                    const std::string& name,
+                                                    ClusterDiscoveryCallbackSharedPtr callback) {
+  ThreadLocalClusterManagerImpl& cluster_manager = *tls_;
+
+  auto [handle, discovery_in_progress] = cluster_manager.cdm_.addCallback(name, callback);
+  if (discovery_in_progress) {
+    // This worker thread has already requested a discovery of a cluster with this name, so nothing
+    // more left to do here.
+    return std::move(handle);
+  }
+  // This seems to be the first request for discovery of this cluster in this worker thread. Rest of
+  // the process may only happen in the main thread.
+  dispatcher_.post([this, odcds_handle = std::move(odcds_handle),
+                    weak_callback = ClusterDiscoveryCallbackWeakPtr(callback), name,
+                    &thread_local_dispatcher = cluster_manager.thread_local_dispatcher_] {
+    // Check for the cluster here too. It might have been added between the time when this closure
+    // was posted and when it is being executed.
+    if (getThreadLocalCluster(name) != nullptr) {
+      if (weak_callback.expired()) {
+        // Not only the cluster was added, but it was also already handled, so don't bother with
+        // posting the callback back to the worker thread.
+        return;
+      }
+      thread_local_dispatcher.post([weak_callback] {
+        if (auto callback = weak_callback.lock(); callback != nullptr) {
+          // If this gets called here, it means that we requested a discovery of a cluster without
+          // checking if that cluster is already known by cluster manager.
```

Comment:
Hm, could be a rather long-winded race, true. Worker thread W1 checks for a cluster that is absent, gets preempted for worker thread W2, which does the same and starts the discovery process. Then main thread sends the gRPC request, receives the reply, warms up the cluster and adds it to the thread local CMs. Then W1 gets a chance to run again, and since it decided that the cluster is missing, it starts the discovery and hit this case in main thread. Long-winded to a point of starvation, so I'd say this is a highly unlikely possibility. But who knows what may happen under the really high load. I'll update the comment.